server.js

```diff
@@ -0,0 +1,386 @@
+const http = require("http");
+const fs = require("fs");
+const credentials = {
+  foo: "bar",
+  user: "pass",
+  guest: "guest"
+};
+
+const server = http.createServer(function(request, response) {
+  // AUTHENTICATION
+
+  if (!request.headers["authorization"]) {
+    response.statusCode = 401;
+    response.setHeader("WWW-Authenticate", 'Basic realm="Secure Area"');
+    response.end("<html><body>Not Authorized</body></html>");
+    return;
+  } else {
+    let buffer = new Buffer(
+      request.headers["authorization"].split(" ")[1],
+      "base64"
+    );
+    let plain_auth = buffer.toString();
+    if (
+      Object.entries(credentials).some(function(value) {
+        return value[0] + ":" + value[1] === plain_auth;
+      }) === true
+    ) {
+      response.statusCode = 200;
+      response.end("<html><body>Credentials accepted</body></html>");
+    } else {
+      response.statusCode = 401;
+      response.setHeader("WWW-Authenticate", 'Basic realm="Secure Area"');
+      response.end("<html><body>Not Authorized</body></html>");
+      return;
+    }
+  }
+
+  // GET REQUESTS
+
+  if (request.method === "GET") {
+    console.log(request.method + " " + request.url);
+    // 404 when request is invalid
+    if (!fs.existsSync("./public" + request.url)) {
+      fs.readFile("./public/404.html", function(error, data) {
+        if (error) {
+          console.log(request.url + " cannot be found");
+        } else {
+          response.writeHead(404, { "Content-type": "text/html" });
+          response.write(data);
+          response.end();
+          console.log(request.url + " cannot be found, here's a 404 instead");
+        }
+      });
+    } else {
+      if (request.url === "/") {
+        // GET /
+        fs.readFile("./public/index.html", function(error, data) {
+          if (error) {
+            console.log(request.url + " cannot be found");
+          } else {
+            response.writeHead(200, { "Content-type": "text/html" });
+            response.write(data);
+            response.end();
+          }
+        });
+        // GET /css/styles.css
+      } else if (request.url === "/css/styles.css") {
+        fs.readFile("./public/css/styles.css", function(error, data) {
+          if (error) {
+            console.log(request.url + " cannot be found");
+          } else {
+            response.writeHead(200, { "Content-type": "text/css" });
+            response.write(data);
+            response.end();
+          }
+        });
+        // GET (request.url)
+      } else if (request.url) {
+        fs.readFile("./public" + request.url, function(error, data) {
+          if (error) {
+            console.log(request.url + " cannot be found");
```

Comment:
Errors should be handled properly, they should send back a response with a appropriate status code and message.